Adds Pair plot

The-Broken-Keyboard opened the pull request:
Adds Pair Plot

**Key Change required in PlotMatrix:**
When we send mask through `map_triangle` ( for divergence ), then it doesn't perform subsetting on `mask` data. Which causes dimension conflict. It was directly adding `**kwargs` to `fun_kwargs`. Now it performs subsetting before adding few keys (  dataArrays or datasets ) of `kwargs` into `fun_kwargs`.

for the following code:

```python
from arviz_base import load_arviz_data

import arviz_plots as azp

azp.style.use("arviz-variat")

data = load_arviz_data("centered_eight")
visuals = {"divergence": True}
pc = azp.plot_pair(
    data,
    backend="bokeh" , 
    var_names=["mu","tau", "theta"],
    coords={"school": ["Choate", "Deerfield"]},
    visuals=visuals,
)
pc.show()
```

we get the following plot:

![image](https://github.com/user-attachments/assets/d5c9a01b-da1b-4345-8e96-0a3433ae75e3)


<!-- readthedocs-preview arviz-plots start -->
----
📚 Documentation preview 📚: https://arviz-plots--287.org.readthedocs.build/en/287/

<!-- readthedocs-preview arviz-plots end -->

The-Broken-Keyboard commented:
@aloctavodia @OriolAbril please review this. I will add documentation and tests once we are done with final version of function

OriolAbril commented on src/arviz_plots/plot_matrix.py:
```diff
@@ -384,7 +384,24 @@ def map_triangle(
-                    fun_kwargs = {**aes_kwargs, **kwargs}
+                    fun_kwargs = {
+                        **aes_kwargs,
+                        **{
+                            key: values
+                            for key, values in kwargs.items()
+                            if not isinstance(values, (xr.DataArray, xr.Dataset))
+                        },
+                        **{
+                            key + "_x": process_kwargs_subset(values, var_name_x, sel_x)
+                            for key, values in kwargs.items()
+                            if isinstance(values, (xr.DataArray, xr.Dataset))
+                        },
+                        **{
+                            key + "_y": process_kwargs_subset(values, var_name_y, sel_y)
+                            for key, values in kwargs.items()
+                            if isinstance(values, (xr.DataArray, xr.Dataset))
+                        },
```
I am a bit more sure after having gone over all the code in the PR. I'll still try to double check at some point today

aloctavodia commented on src/arviz_plots/plots/pair_plot.py:
```diff
@@ -0,0 +1,222 @@
+    visuals : mapping of {str : mapping or False}, optional
+        Valid keys are:
+
+        * scatter -> passed to :func:`~.visuals.scatter_couple`
+        * divergence -> passed to :func:`~.visuals.scatter_couple`. Defaults to False.
+        * label -> :func:`~.visuals.label_plot`
```
ohh, I assumed that was because of some limitation of the current PlotMatrix implementation. We definitely want to show axes by default.

The-Broken-Keyboard commented:
@OriolAbril @aloctavodia 

I have incorporated all the changes suggested above

**MAJOR CHANGES**
- Added a `map_row_col` method which takes a variable, selection dict and orientation (row or col) and then maps the given function to whole row or col of matrix. Since we have this method in PlotMatrix, so we can also include similar kind of method  in PlotCollection and the work around implemented in #208 will not be needed now, because we can use this function in plotcollection to handle conflicting xlabels issue too. So, i removed the work around of #208 for now ( to incorporate this new function ). 
- Added parameters `triangle` and `remove_axis` ( in visuals ) .
- Added marginals ( supports kind="kde","hist" and "ecdf") on diagonal. 
- Also added diagonal xlabels if `triangle="upper"` is given
- Diagonal labels will be used automatically if `marginal=False`
- x labels for bottom plots and y labels for left plots will be used only if `marginal=True`, because otherwise diagonal labels will be there to label the rows and cols(automatically, as mentioned above).
- **Enforcing `remove_axis=False` if `marginal=True`. Because if we have marginals we definitely need xlabels and ylabels to label rows and cols ( since diagonal labels won't be there and removing axis will remove axis labels too).
- `sharex` and `sharey` is set to `True`. If `marginal=True` then we can't have `sharey=True`, so that is forced to be `False`.

** We are getting warning in Bokeh whenever we use upper or lower triangle, it says renderer wasn't found for some plots**

Examples:

### Code 1

```python
from arviz_base import load_arviz_data

import arviz_plots as azp

azp.style.use("arviz-variat")

data = load_arviz_data("centered_eight")
pc = azp.plot_pair(
    data,
    backend="bokeh" , 
    var_names=["mu","tau","theta"],
    coords={"school": ["Choate", "Deerfield"]},
    marginal=True,
    marginal_kind="kde",
    triangle="both",
    visuals={"divergence": True},
)
pc.show()
```
### result 1
![image](https://github.com/user-attachments/assets/5f8d51c5-6834-492c-943a-8190dd774625)

### code 2

```python
from arviz_base import load_arviz_data

import arviz_plots as azp

azp.style.use("arviz-variat")

data = load_arviz_data("centered_eight")
pc = azp.plot_pair(
    data,
    backend="bokeh" , 
    var_names=["mu","tau","theta"],
    coords={"school": ["Choate", "Deerfield"]},
    marginal=True,
    marginal_kind="hist",
    triangle="both",
    visuals={"divergence": True},
)
pc.show()
```

### result 2
![image](https://github.com/user-attachments/assets/3f97f8f1-fa4f-471c-8d8e-14c98d972056)

### code 3

```python
from arviz_base import load_arviz_data

import arviz_plots as azp

azp.style.use("arviz-variat")

data = load_arviz_data("centered_eight")
pc = azp.plot_pair(
    data,
    backend="bokeh" , 
    var_names=["mu","tau","theta"],
    coords={"school": ["Choate", "Deerfield"]},
    marginal=True,
    marginal_kind="ecdf",
    triangle="both",
    visuals={"divergence": True},
)
pc.show()
```

### result 3
![image](https://github.com/user-attachments/assets/852905c7-5e2c-48fa-8ee7-1c3b09189d6f)

### code 4

```python
from arviz_base import load_arviz_data

import arviz_plots as azp

azp.style.use("arviz-variat")

data = load_arviz_data("centered_eight")
pc = azp.plot_pair(
    data,
    backend="bokeh" , 
    var_names=["mu","tau","theta"],
    coords={"school": ["Choate", "Deerfield"]},
    marginal=True,
    marginal_kind="kde",
    triangle="upper",
    visuals={"divergence": True},
)
pc.show()
```

![image](https://github.com/user-attachments/assets/e49a0091-96a9-4f81-89a0-3ccca6b540ad)

### code 5

```python
from arviz_base import load_arviz_data

import arviz_plots as azp

azp.style.use("arviz-variat")

data = load_arviz_data("centered_eight")
pc = azp.plot_pair(
    data,
    backend="bokeh" , 
    var_names=["mu","tau","theta"],
    coords={"school": ["Choate", "Deerfield"]},
    marginal=True,
    marginal_kind="kde",
    triangle="lower",
    visuals={"divergence": True},
)
pc.show()
```

### result 5
![image](https://github.com/user-attachments/assets/1dc18f24-6716-41d7-895e-77b4767f89f3)

### code 6
```python
from arviz_base import load_arviz_data

import arviz_plots as azp

azp.style.use("arviz-variat")

data = load_arviz_data("centered_eight")
pc = azp.plot_pair(
    data,
    backend="bokeh" , 
    var_names=["mu","tau","theta"],
    coords={"school": ["Choate", "Deerfield"]},
    marginal=False,
    triangle="both",
    visuals={"divergence": True, "remove_axis":"both"},
)
pc.show()
```
### result 6
![image](https://github.com/user-attachments/assets/6c1f7ead-acf7-44f7-9d49-93402d4d31a0)

The-Broken-Keyboard commented on src/arviz_plots/plots/pair_plot.py:
```diff
@@ -0,0 +1,395 @@
+        pc_kwargs["figure_kwargs"].setdefault("sharey", True)
+        pc_kwargs["figure_kwargs"].setdefault("sharex", True)
+        if marginal:
+            pc_kwargs["figure_kwargs"]["sharey"] = False
```
This required a little bit logic change in bokeh backend's create_plotting_grid function. Before the logic was such that we couldn't have sharex="col" and sharey="row" simultaneously. Since they were mutually exclusive so it was throwing error . Fixed that in backend.

Also since having just text on diagonal plots when marginal is False was little bit weird because we required to give x and y location of text which varied for different variables ( in case if we want them to be at centre). So now I added scatter plot for diagonal plots as well when marginal is False. Otherwise we will have marginals

OriolAbril replied:
> This required a little bit logic change in bokeh backend's create_plotting_grid function

Thanks! Bokeh has always been the one to generate more trouble for us so it ends up being less used and worked on. The refactor is making things much better on this end but still not fully there.

> Also since having just text on diagonal plots when marginal is False was little bit weird because we required to give x and y location of text which varied for different variables ( in case if we want them to be at centre). So now I added scatter plot for diagonal plots as well when marginal is False. Otherwise we will have marginals

I don't think there is any case where we want the scatter against itself. It is ok as a subproduct in `plot_pair_focus` when removing it grealy increases complexity of the code, but I would not add it explicitly.

I see two options on this. One is doing some updates to the initialization of PlotMatrix in order to request a grid with one less column but same number of rows, and storing it like so in the "plot" dataarray (which would still have the same shape and coordinates) with `None` on the diagonal. Example for 3 variables, we generate a 3x2 grid:

```
None | 0, 0 | 0, 1
1, 0 | None | 1, 1
2, 0 | 2, 1 | None
```

This means there are no diagonal elements to be filled at all.

Alternatively, adapt the code for the labels along the diagonal. The backend text function takes vertical and horizontal alignment which we can set to center the text around the provided `x` and `y`. And as that would be for the case without marginals with axis sharing, the x and y would actually be the same value, and we can compute it as:

```
text_center = (distribution.max(dim=sample_dims) + distribution.min(dim=sample_dims)) / 2
```

which should get something that looks centered

OriolAbril commented on src/arviz_plots/plot_matrix.py:
```diff
@@ -174,10 +218,29 @@ def get_target(self, var_name, selection, var_name_y=None, selection_y=None):
+            var_name_y = var_name
+            selection_x = self.fixed_selection
+            selection_y = selection
+        else:
+            var_name_x = var_name
+            selection_x = selection
```
Given the behaviour of `subset_matrix_da` I think this means that the simple fact of `var_name_y` and `selection_y` being None will already behave like "diagonal" even if the `orientation` attribute is not None. I think it is actually fine and we only need to check orientation `row` or `col`, with no orientation meaning "diagonal" which I think is a good default.

OriolAbril commented on src/arviz_plots/plot_matrix.py:
```diff
@@ -355,6 +418,9 @@ def map_triangle(
                 ignore_aes=ignore_aes,
             )
+
+        self.orientation = None
```
I would actually `raise` if `orientation` is set here (see other comments too).

OriolAbril commented on src/arviz_plots/plot_matrix.py:
```diff
@@ -158,6 +160,48 @@ def _generate_viz_dt(self, **figure_kwargs):
+    def orientation(self):
+        """Orientation of targets we want to fetch through :meth:`~PlotMatrix.get_target`.
+
+        Orientation is always taken into account when using
+        :meth:`~PlotMatrix.get_target` to fetch target for plotting.
+        """
+        return self._orientation
+
+    @orientation.setter
+    def orientation(self, value):
+        self._orientation = value
+
+    @property
+    def fixed_var_name(self):
+        """Fixed variable's name of targets we fetch through :meth:`~PlotMatrix.get_target`.
+
+        fixed_var_name is taken into account when using :meth:`~PlotMatrix.get_target`
+        to fetch target for plotting and :attr:`~PlotMatrix.orientation` is
+        set to either `row` or `col`.
+
+        It tells the fixed variable name in a `row` or `col`
+        """
+        return self._fixed_var_name
+
+    @fixed_var_name.setter
+    def fixed_var_name(self, value):
+        self._fixed_var_name = value
+
+    @property
+    def fixed_selection(self):
+        """Fixed dictionary for subsetting of targets."""
+        return self._fixed_selection
+
+    @fixed_selection.setter
+    def fixed_selection(self, value):
+        self._fixed_selection = value
+
+    @fixed_var_name.setter
+    def fixed_var_name(self, value):
+        self._fixed_var_name = value
```
I think we could have them as "simple" attributes set to `None` in `__init__`. Moreover, I don't think we expect/want users to interact with them directly but more so as an internal thing to reduce duplication between `map` and `map_row_col`

OriolAbril commented on src/arviz_plots/plot_matrix.py:
```diff
@@ -439,10 +510,98 @@ def map(
+            store_artist=store_artist,
+            artist_dims=artist_dims,
+            **kwargs,
+        )
```
I would revert `orientation` back to `None` before returning.

OriolAbril commented on src/arviz_plots/plot_matrix.py:
```diff
@@ -473,11 +632,11 @@ def map(
         --------
         arviz_plots.PlotMatrix.map_triangle
         """
+        self.set_fixed_var_attributes(index, orientation)
```
Here also revert the 3 attributes to `None` before returning.

OriolAbril reviewed (commented):
next round of review. I think it is shaping up very well

OriolAbril commented on src/arviz_plots/plots/pair_plot.py:
```diff
@@ -0,0 +1,452 @@
+    coords=None,
+    sample_dims=None,
+    marginal=True,
+    marginal_kind="kde",
```
default to rcparam or even none and pass to plot_dist so plot_dist changes none to the rcparam

OriolAbril commented on docs/source/gallery/distribution/08_plot_pair_distribution.py:
```diff
@@ -0,0 +1,27 @@
+    data,
+    var_names=["mu","theta","tau"],
+    coords= {"school": ["Choate", "Deerfield"]},
+    marginal=True,
+    marginal_kind="kde",
```
I would remove these two arguments given they are the defaults

OriolAbril commented on docs/source/gallery/inference_diagnostics/10_plot_pair.py:
```diff
@@ -0,0 +1,29 @@
+dt = load_arviz_data("centered_eight")
+pc = azp.plot_pair(
+    dt,
+    var_names=["theta", "mu"],
```
```suggestion
    var_names=["theta", "tau"],
```

Tau is the interesting one to vizualize with divergences

OriolAbril commented on docs/source/gallery/inference_diagnostics/10_plot_pair.py:
```diff
@@ -0,0 +1,29 @@
+    var_names=["theta", "mu"],
+    coords={"school": ["Lawrenceville", "Mt. Hermon"]},
+    visuals={"divergence":True},
+    marginal=True,
+    marginal_kind="kde",
```
maybe change this so the plot looks more different than the one in the distributions section

```suggestion
    marginal=False,
```

OriolAbril commented on src/arviz_plots/backend/bokeh/__init__.py:
```diff
@@ -317,31 +317,33 @@ def create_plotting_grid(
             if width_ratios is not None:
                 subplot_kws["width"] = plot_widths[col]
+
+            if row * cols + (col + 1) > number:
```
we should add tests for this, but we don't yet have the setup for backend specific tests. I will ping you once I start working on [this issue](https://github.com/arviz-devs/arviz-plots/issues/100) so we can both add tests for backend functions

The-Broken-Keyboard replied:
sure

OriolAbril commented on src/arviz_plots/plot_matrix.py:
```diff
@@ -439,10 +459,172 @@ def map(
+            Dictionary of {coordinate names : coordinate values} that should
+            be used to subset the aes, data and viz objects before any faceting
+            or aesthetics mapping is applied.
+        ignore_aes : set, optional
```
```suggestion
        ignore_aes : str or set of str, default "all"
```

OriolAbril commented on src/arviz_plots/plot_matrix.py:
```diff
@@ -439,10 +459,172 @@ def map(
+        artist_dims : mapping of {hashable : int}, optional
+            Dictionary of sizes for proper allocation and storage when using
+            ``map`` with functions that return an array of :term:`visual`.
+        **kwargs : mapping, optional
```
```suggestion
        **kwargs
```

OriolAbril commented on src/arviz_plots/plot_matrix.py:
```diff
@@ -439,10 +459,172 @@ def map(
+
+        See Also
+        --------
+        arviz_plots.PlotMatrix.map_triangle
```
I think all 4: map, map_triangle, map_row and map_col should reference the other three in their see also sections, with the first one being the most relevant one, map_row<->map_col and map<->map_triangle.

OriolAbril commented on tests/test_plots.py:
```diff
@@ -332,6 +333,59 @@ def test_plot_mcse_models(self, datatree, datatree2, backend):
         assert "model" in pc.aes["color"].dims
         assert "/x" in pc.aes.groups
 
+    def test_plot_pair(self, datatree, backend):
```
either here or in follow up tests we should make sure that `triangle` argument is taken into account. That is, with triangle both `scatter` only has `None` on the diagonal, with lower on the diagonal and above... Same thing for `marginal=False`

OriolAbril commented on docs/source/gallery/distribution/08_plot_pair_distribution.py:
```diff
@@ -0,0 +1,27 @@
+"""
+# Scatterplot all variable against each other
```
```suggestion
# Scatterplot all variables against each other
```

OriolAbril reviewed (commented):
mostly test and doc polishing left and we can merge

OriolAbril commented on src/arviz_plots/plots/pair_plot.py:
```diff
@@ -0,0 +1,495 @@
+        )
+
+    # default removal of axis for better visualization
+    default_remove_axis_kwargs = {}
+    default_remove_axis_kwargs["axis"] = "both"
```
There is nothing where we are using this as dict, it is much easier to follow the code if instead you replace the `**default_remove_axis_kwargs` for `axis="both"`.

OriolAbril commented on src/arviz_plots/plots/pair_plot.py:
```diff
@@ -0,0 +1,495 @@
+    -------
+    PlotMatrix
+
+    Examples
+    --------
+    Default plot_pair
```
The examples should focus on illustrating the behaviour that is specific to `plot_matrix` and can be inferred/learned from the general tutorial. At the same time doing that helps clarify the behaviour of unique/complicated arguments.

In this case I would generate multiple plots with `plot_pair(dt, var_names=["mu", "tau"], marginal=, triangle=)` for different values of marginal and triangle to show the less common/potentially surprising cases. For example, with lower triangle only, the behaviour of `marginal` True/False is quite simple to explain, if the diagonal is free the labels with variable names go there, otherwise they go at the left/bottom of the figure. However, the cases with upper are not as polished (which is ok) but should have extensive docs to minimize issues and surprises

OriolAbril commented on src/arviz_plots/plots/pair_plot.py:
```diff
@@ -0,0 +1,495 @@
+        else:
+            backend = plot_matrix.backend
+
+    if marginal_kind is None:
+        marginal_kind = rcParams["plot.density_kind"]
```
This is not used anywhere, so we can leave the substitution of None for the rcParam to `plot_dist`

OriolAbril commented on src/arviz_plots/plots/pair_plot.py:
```diff
@@ -0,0 +1,495 @@
+        dist_plot_stats["point_estimate"] = stats.get("point_estimate", {})
+
+        plot_matrix = plot_dist(
+            dt,
+            var_names,
+            filter_vars,
+            group,
+            coords,
+            sample_dims,
+            marginal_kind,
```
```suggestion
            distribution,
            sample_dims=sample_dims,
            kind=marginal_kind,
```

We have already done all the filtering/subsetting to get `distribution` and `plot_dist` doesn't need data from any other group, so we can pass distribution to avoid processing the same thing twice.
